mpi4jax/_src/tokenizer.py

```diff
@@ -0,0 +1,186 @@
+import jax
+from jax import linear_util as lu
+from jax.interpreters import xla
+
+
+token_override_registry = {}  # Dict[Callable, Callable]
+recursive_token_forwarding_registry = {}  # Dict[Primitive, Callable]
+
+
+def safe_map(f, *args):
+    args = [list(arg) for arg in args]
+    for arg in args:
+        assert len(arg) == len(args[0]), f"Mismatched lengths: {[len(x) for x in args]}"
+    return [f(*arg) for arg in zip(*args)]
+
+
+def xla_call_overrride(read, eqn, token):
+    subfuns, bind_params = eqn.primitive.get_bind_params(eqn.params)
+    bind_params["donated_invars"] = (False,) + bind_params["donated_invars"]
+    map_token = lambda func: lu.wrap_init(  # noqa: E731
+        lambda token, *args: _token_forwarding(func.call_wrapped, token)(*args)
+    )
+    subfuns = safe_map(map_token, subfuns)
+    ans = eqn.primitive.bind(
+        *subfuns, token, *safe_map(read, eqn.invars), **bind_params
+    )
+    token = ans[0]
+    ans = ans[1:]  # Drop the token.
+    return token, ans
+
+
+recursive_token_forwarding_registry[xla.xla_call_p] = xla_call_overrride
+
+
+def scan_override(read, eqn, token):
+    _, bind_params = eqn.primitive.get_bind_params(eqn.params)
+    new_body_fn = lambda token, *args: _token_forwarding(  # noqa: E731
+        jax.core.jaxpr_as_fun(bind_params["jaxpr"]), token
+    )(*args)
+    bind_params["jaxpr"] = jax.make_jaxpr(new_body_fn)(
+        token, *safe_map(read, eqn.invars)
+    )
+    # Update bind_params to account for the additional token.
+    bind_params["num_carry"] += 1
+    bind_params["linear"] = (False,) + bind_params["linear"]
+    ans = eqn.primitive.bind(token, *safe_map(read, eqn.invars), **bind_params)
+    token = ans[0]
+    ans = ans[1:]  # Drop the token.
+    return token, ans
+
+
+recursive_token_forwarding_registry[jax.lax.scan_p] = scan_override
+
+
+def while_override(read, eqn, token):
```

Comment:
Does this also cover `fori_loop`?

Reply:
`fori_loop` is handled by the `scan_override`. We test `fori_loop` in `test_fori_loop_tokenizer` in `tests/test_auto_tokenizer.py`